drivers: w1: Add DS2485 1-Wire master driver

str4t0m opened the pull request:
This PR adds a driver for the DS2485 1-Wire master

str4t0m committed:
drivers/w1: driver for ds2485 1-Wire master

This commit introduces the 1-wire master driver for maxim ds2485.

The ds2485 master has nearly the same (1-wire) feature set and
i2c-interface as the ds2477.
Therefore the common parts are extracted, but to avoid
any nda troubles only the ds2485 specific part is included.

Compared to older 1-wire masters, the ds2485 supports higher level
commands, supporting multi byte operations, search next, automatic crc
calculation.

In this driver only basic read and write operations are supported,
further hardware features are not yet utilized by the driver.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
samples/drivers/w1: scanner: add ds2485 overlay

Adds an overlay for the ds2485 and adds a test configuration
for it.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
tests/drivers/w1: add ds2485 overlay and configuration

Adds the ds2485 driver overlay and a test configuration
for it.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m committed:
dts/bindings: w1 move active-pullup property to w1-master

This property is supported by most 1-Wire masters,
move it to the w1-master binding.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

galak commented on dts/bindings/w1/maxim,ds2477_85_common.yaml:
```diff
@@ -0,0 +1,93 @@
+        "333"
+
+  standard-slew:
+    type: string
```
should be type `int`

galak commented on dts/bindings/w1/maxim,ds2477_85_common.yaml:
```diff
@@ -0,0 +1,93 @@
+        "1300"  1300ns
+
+  overdrive-slew:
+    type: string
```
should be type int

galak reviewed (dismissed):
dt-binding change.

str4t0m commented:
Thanks for the first review.
- Updated the called out properties to int and changed the initialization to to using `DT_ENUM_IDX` for all enums.
- Additionally using `select I2C` instead of `depends on I2C`

ycsin commented on drivers/w1/w1_ds2477_85_common.c:
```diff
@@ -0,0 +1,294 @@
+	k_usleep(cfg->t_op_us);
+
+	ret = i2c_read_dt(&cfg->i2c_spec, buf, 2);
+	if (ret != 0 || buf[0] != 1 || buf[1] != RESULT_SUCCESS) {
```
To be more MISRA-ish

```suggestion
	if ((ret != 0) || (buf[0] != 1) || (buf[1] != RESULT_SUCCESS)) {
```

ycsin replied:
Also, should it be `ret < 0` instead?

str4t0m replied:
Changed these comparisons accordingly.

ycsin commented on drivers/w1/w1_ds2477_85_common.c:
```diff
@@ -0,0 +1,294 @@
+	int w1_timing = cfg->mode_timing[data->master_reg.od_active].t_slot * 8 * tx_len;
+	uint8_t buf[3] = {CMD_WRITE_BLOCK, (tx_len + 1), 0};
+	struct i2c_msg tx_msg[2] = {
+		{.buf = buf, .len = 3, .flags = I2C_MSG_WRITE},
+		{.buf = (uint8_t *)buffer, .len = tx_len, .flags = (I2C_MSG_WRITE | I2C_MSG_STOP)}
```
nits

```suggestion
		[0] = {.buf = buf, .len = 3, .flags = I2C_MSG_WRITE},
		[1] = {.buf = (uint8_t *)buffer, .len = tx_len, .flags = (I2C_MSG_WRITE | I2C_MSG_STOP)},
```

str4t0m replied:
Adding the index would bring the line past 100 chars, so I believe it wouldn't get more readable.
So if it's just a preference I'd leave, otherwise I'll change it of course.

ycsin commented on drivers/w1/w1_ds2477_85_common.c:
```diff
@@ -0,0 +1,294 @@
+	struct w1_ds2477_85_data *data = dev->data;
+
+	switch (type) {
+	case W1_SETTING_SPEED:
```
Is it expected to have mroe cases in the future? Otherwise I think if-else is more intuitive?

str4t0m replied:
Yes, there are definitely coming more, I just didn't include things like strong-pull ups yet because I'd like to check a few things first such that different master can do it in a uniform way.

ycsin commented on drivers/w1/w1_ds2477_85_common.h:
```diff
@@ -0,0 +1,258 @@
+#define CMD_COMPUTE_CRC		 0xCC
+
+/* I2C communication result bytes */
+#define RESULT_SUCCESS	     0xAA
```
Would be better if these are namespaced so that they wont conflict with other stuff

str4t0m replied:
Prefixed these results bytes with, DS24.
Such that a conflict is unlikely, but the names are still not too long (i.e. DS24_RES_SUCCESS)

str4t0m commented:
Rebased on latest main.

dp7hgh7 commented on drivers/w1/w1_ds2477_85_common.c:
```diff
@@ -0,0 +1,294 @@
+int ds2477_85_write_port_config(const struct device *dev, uint8_t reg, uint16_t value)
+{
+	const struct w1_ds2477_85_config *cfg = dev->config;
+	uint8_t buf[5] = {CMD_WRITE_W1_PORT_CONFIG, 3, reg};
```
Can you give those magic numbers names? It's not directly obvious what they mean. Goes for multiple occasions.

str4t0m replied:
These magic numbers indicate the command length.
I'll abstract these into `CMD_xyz_LEN` and similar for all the I2C commands.
I believe I can pass the rx/tx-len of the script commands (`cfg->w1_script_cmd()`) directly, or would you abstract these as well?

dp7hgh7 replied:
"ARRAY_SIZE" would be perfect but no super strong opinion on that

str4t0m commented:
Could you please have another look.

ycsin commented on drivers/w1/w1_ds2477_85_common.c:
```diff
@@ -0,0 +1,294 @@
+
+	ret = i2c_read_dt(&cfg->i2c_spec, buf, 2);
+	if ((ret < 0) || (buf[0] != 1) || (buf[1] != DS2477_88_RES_SUCCESS)) {
+		return -EIO;
```
Just wondering if it is possible to return `ret` here instead of `-EIO`?

str4t0m replied:
I'd have to introduce a second `if()` for the buffer checks, and wouldn't change much because I2c_read is expected to return either `0` or `-EIO`.

str4t0m replied:
If now changed this and all similar occurrences as well.

dp7hgh7 replied:
I think this is the wrong order. Currently it first checks the content and afterwards if the transaction was successful. Not dramatic in this case but might shadow the actual error from the I2C transaction.

str4t0m replied:
I have now restored the order. Thanks.

ycsin commented on drivers/w1/w1_ds2477_85_common.c:
```diff
@@ -0,0 +1,294 @@
+		return -EIO;
+	}
+
+	return ret;
```
Inconsistent with L66

ycsin commented on drivers/w1/w1_ds2477_85_common.h:
```diff
@@ -0,0 +1,279 @@
+#define PORT_REG_STANDARD_SPEED_T_MSR	0x07
+#define PORT_REG_STANDARD_SPEED_T_REC	0x08
+#define PORT_REG_OVERDRIVE_SPEED_T_RSTL 0x09
+#define PORT_REG_OVERDRIVE_SPEED_T_MSI	0x0a
```
The defines before these were using capitalized letters

str4t0m replied:
now all uniformly use lower-case.

ycsin commented on drivers/w1/w1_ds2485.c:
```diff
@@ -0,0 +1,115 @@
+	}
+	k_usleep(DS2485_T_OSCWUP_us);
+
+	ret = w1_ds2477_85_init(dev);
```
I think it is same as

```c
return w1_ds2477_85_init(dev);
```

?

str4t0m commented:
@galak @ycsin I have addressed your comments, could you please have another look.
